core/task.h

```diff
@@ -33,55 +33,150 @@
 
 #include <string>
 
+#include "gate.h"
+#include "pktbatch.h"
+#include "utils/extended_priority_queue.h"
+
 struct task_result {
   bool block;
   uint32_t packets;
   uint64_t bits;
 };
 
+typedef uint16_t gate_idx_t;
 typedef uint16_t task_id_t;
 typedef uint64_t placement_constraint;
 
+#define MAX_PBATCH_CNT 256
+
 class Module;
 
 namespace bess {
 class LeafTrafficClass;
+class PacketBatch;
 }  // namespace bess
 
 // Functor used by a leaf in a Worker's Scheduler to run a task in a module.
 class Task {
+ private:
+  // Used by operator().
+  Module *module_;
+  void *arg_;                  // Auxiliary value passed to Module::RunTask().
+  bess::LeafTrafficClass *c_;  // Leaf TC associated with this task.
+
+  struct GateBatchGreater {
+    bool operator()(std::pair<bess::IGate *, bess::PacketBatch *> left,
+                    std::pair<bess::IGate *, bess::PacketBatch *> right) const {
+      return left.first->priority() > right.first->priority();
+    }
+  };
+
+  // XXX Tasks needs to be non-const in workers/modules
+  mutable bess::utils::extended_priority_queue<
+      std::pair<bess::IGate *, bess::PacketBatch *>, GateBatchGreater>
+      igates_to_run_;  // A queue for IGates to run
+
+  mutable bess::IGate *next_gate_;  // Cache next module to run without merging
+  // Optimization for chain
+  mutable bess::PacketBatch
+      *next_batch_;  // Cache to run next batch with next module
+
+  mutable bess::PacketBatch
+      dead_batch_;  // A packet batch for storing packets to free
+
+  // Simple packet batch pool
+  mutable int pbatch_idx_;
+  mutable bess::PacketBatch *pbatch_;
+
+  mutable std::vector<bess::PacketBatch *> gate_batch_;
+
+  /* The current input gate index is not given as a function parameter.
+   * Modules should use get_igate() for access */
+  mutable gate_idx_t current_igate_;
+
  public:
   // When this task is scheduled it will execute 'm' with 'arg'.  When the
   // associated leaf is created/destroyed, 'module_task' will be updated.
-  Task(Module *m, void *arg) : module_(m), arg_(arg), c_(nullptr) {}
+  Task(Module *m, void *arg)
+      : module_(m),
+        arg_(arg),
+        c_(nullptr),
+        igates_to_run_(),
+        next_gate_(),
+        next_batch_(),
+        pbatch_idx_(),
+        pbatch_(
+            new bess::PacketBatch[MAX_PBATCH_CNT]),  // XXX Need to adjust size
+        gate_batch_(std::vector<bess::PacketBatch *>(64, 0)) {
```

Comment:
Could you add `current_igate_` (and in general all members) to the initializer list?